base/compiler/optimize.jl

```diff
@@ -32,14 +32,15 @@ function default_inlining_policy(@nospecialize(src))
     if isa(src, CodeInfo) || isa(src, Vector{UInt8})
         src_inferred = ccall(:jl_ir_flag_inferred, Bool, (Any,), src)
         src_inlineable = ccall(:jl_ir_flag_inlineable, Bool, (Any,), src)
-        return src_inferred && src_inlineable
+        return src_inferred && src_inlineable ? src : nothing
     end
-    return false
+    return nothing
 end
 
 mutable struct OptimizationState
     linfo::MethodInstance
     src::CodeInfo
+    ir::Any # Union{Nothing, IRCode}
```

Comment:
having both, and partly nullable, seems non-ideal. Can we promote a `CodeInfo` early?

Reply:
Unfortunately we need to keep the CodeInfo around, because it has some metadata and flags that we don't keep elsewhere and our replacement procedure assumes we have the old CodeInfo. We should clean this up in the fullness of time, but this seemed easiest for the moment.

Reply:
ah, I guess at some point we should make IRCode a true superset, but this is okay for now in that case